Product Collection: Add columns control to product collection block editor settings

imanish003 opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds Columns panel in sidebar settings of Product Collection block. 
<img width="279" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/7818b070-37f8-4adb-96d9-dccc5f75851b">

<!-- Reference any related issues or PRs here -->

Fixes #9356 

### Testing

1. Create a new post
2. Add Product Collection block
3. Verify that the sidebar displays a Columns setting when the Product Collection block is selected.
   - Ensure that the Columns setting is clearly visible and accessible in the sidebar.
   - Confirm that the Columns setting allows for selecting different column options, such as 2, 3, or 4 columns.
<img width="279" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/7818b070-37f8-4adb-96d9-dccc5f75851b">


4. Change the number of columns using the Columns setting.
   - Confirm that the change is immediately reflected in the editor.
   - For example, if you set 4 columns using the new control, ensure that the products are displayed in 4 columns within the editor.
5. Save and publish the post.
   - Verify that the same change is reflected on the frontend.
   - Check that the product collection block displays the same number of columns as set in the editor.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/9219776c-e8d2-4e43-b7a9-9078e42af23a



* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add columns control to product collection block editor settings

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/columns-control.tsx:
```diff
@@ -0,0 +1,39 @@
+			min={ 2 }
+			max={ Math.max( 6, columns ) }
+		/>
+	) : (
+		<></>
+	);
```
What about returning `null` instead which I think is a bit clearer (but it's opinionated, so feel free to ignore the comment)

imanish003 replied:
Actually, I tried using the following code, but I was getting a TypeScript error. Therefore, I decided to return `<></>` instead:
```
	return (
		showColumnsControl && (
			<RangeControl
				label={ __( 'Columns', 'woo-gutenberg-products-block' ) }
				value={ columns }
				onChange={ ( value: number ) =>
					props.setAttributes( {
						displayLayout: {
							...props.attributes.displayLayout,
							columns: value,
						},
					} )
				}
				min={ 2 }
				max={ Math.max( 6, columns ) }
			/>
		)
	);
```

However, you are right, I can also return null. I don't get a TypeScript error if I return null. I will make the change. Thanks. 🤝

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/columns-control.tsx:
```diff
@@ -0,0 +1,39 @@
+	props: BlockEditProps< ProductCollectionAttributes >
+) => {
+	const { type, columns } = props.attributes.displayLayout;
+	const showColumnsControl = type === 'flex';
```
What are other types available? Is it possible to change it? Or it will be controlled by patterns and in some cases users won't be able to change columns (for example if we display products in rows?)

imanish003 replied:
As you can see in the Products beta block, there is a List view and Grid view toolbar:

<img width="837" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/29a6d685-f916-44bc-9b8b-8e000c9c32c3">

Therefore, we will have two display types. We can implement the same toolbar, allowing users to change the type. If we don't have an existing issue for this, I think we should create a new one.

kmanijak reviewed (approved):
Works as expected! I left just minor comments 💪

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns
